utils/stall_free.hh

```diff
@@ -56,14 +58,93 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+template <typename T>
+concept GentlyClearable = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.reset() } -> std::same_as<void>;
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept VectorLike = requires (T x, size_t n) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept Container = !VectorLike<T> && requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.erase(x.begin()) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <Container T>
+future<> clear_gently(T& c) noexcept;
+
+template <GentlyClearable T>
+future<> clear_gently(T& o) noexcept(noexcept(o.clear_gently())) {
+    return o.clear_gently();
+}
+
+template <typename T>
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
+
+template <SmartPointer Ptr>
+future<> clear_gently(Ptr& o) {
+    return clear_gently(*o).finally([&o] {
+        return o.reset();
+    });
+}
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept {
+    return do_for_each(a, [] (T& o) {
+        return clear_gently(o);
+    });
+}
+
+// Trivially destructible elements can be safely cleared in bulk
+template <VectorLike T>
+requires std::is_trivially_destructible_v<typename T::value_type>
+future<> clear_gently(T& v) noexcept {
+    v.clear();
+    return make_ready_future<>();
+}
+
+// Clear the elements gently and destroy them one-by-one
+// in reverse order, to avoid copying.
+template <VectorLike T>
+requires (!std::is_trivially_destructible_v<typename T::value_type>)
+future<> clear_gently(T& v) noexcept {
+    return do_until([&v] { return v.empty(); }, [&v] {
+        return clear_gently(v.back()).finally([&v] {
+            v.pop_back();
+        });
+    });
+    return make_ready_future<>();
+}
+
+template <typename K, typename T>
+future<> clear_gently(std::pair<const K, T>& p) {
+    return clear_gently(p.second);
+}
```

Comment:
I don't understand this implementation. It seems completely arbitrary. I'm guessing it was designed for maps or something - but and can be misused by accidentally give it a pair that has nothing to do with maps.
Could we make all these auxillary clear_gently functions that end-users should not use hidden? By the way, can't/shouldn't we move most of this code to a .cc file, and only keep some of it in the header file?

Reply:
To your first question - yes, it's intended to be used internally as a helper function to the loop that iterates an associative container.  I'll see if I can come up with a clean way to hiding these helpers from the user.

To move it to a source file we'll have to instantiate all the supported the specializations, no?
Regardless if they are used or not by the application.  I'm not sure that's a good idea.

Reply:
Please note that it's more than just "hiding" these weird implementations for pair (clearing the second) and misc (do absolutely nothing) by giving them a strange name. You also don't want the top-level clear_gently() implementation to not be able to work on these cases. You want clear_gently() to support only types that match specific concepts - I think it should fail to compile with other types, such as a pair.

Reply:
If you give them a new name, the whole thing falls apart. This is a variation of a destructor. Like a destructor, it has to destroy everything recursively. If some objects have their clear_gently destructors named differently, recursion breaks.

Reply:
It doesn't have to be entirely recursive until the very bottom. You can support specific kind of leaves in a special way, e.g., a map-like object (however we define this...) will have its values cleared but not keys (I guess you assume that keys are small objects??) - you don't need to continue recursing all the way down to an std::pair and then define an odd way to clear it, so some default implementation that does nothing for most types.
If the implementation for std::pair cleared both items, and if we had a default implementation that does something reasonable (what?) that's one thing. But having implementations that do strange things for std::pair and unknown types without any warnings whatsoever - I think that's dangerous.

Reply:
You can stop the recursion earlier, but the code bloats. It means that every implementation will have to do something like

```c++
    if constexpr (has_clear_gently<T>) {
         return x.clear_gently();
    } else {
         return make_ready_future<>();
    }
```

Do we want this to be repeated everywhere? I think not.

Reply:
Whatever we do, I don't want a year from now a bug where we did

     something.clear_gently()

And this clear_gently() simply didn't do anything and returned a ready future. I think it's fine that clear_gently() only works on specific types we designed it for, and it doesn't need to compile and return ready_future for other things we don't recognize.

One final idea: clear_gently() is supposed to be a futurized version of clear(). So we should define it **only** on types which have clear(). clear_gently's implementation will use internal::do_clear_gently() whose implementation will be recursive, but users will not be able to use it directly.

There is one last risk of a problem: imagine you have an std::list<std::pair<std::list, std::list>>> - our clear_gently will decend into the list, find a pair, and thinking it's a map (?) will clear_gently just the second list in the pair, and destroy (not gently) the first one. This is probably wrong - even though the example is contrived.

Reply:
Let's say we call clear_gently(foo), where foo has type std::list<T>. Should it fail if T doesn't have clear_gently()?

If it does, then we can't clear_gently if T==int or some random struct.

If it doesn't, then we can accidentally bypass the protection you want by clearing std::list<T>, but not when clearing T itself.

Reply:
In the next version I'm working on I defined the following:
```c++
namespace internal {

template <typename T>
concept HasClearGentlyImpl = requires (T x) {
    { clear_gently(x) } -> std::same_as<future<>>;
};

template <typename T>
requires HasClearGentlyImpl<T>
future<> clear_gently(T& x) noexcept {
    return utils::clear_gently(x);
}

// This default implementation of clear_gently
// is required to "terminate" recursive clear_gently calls
// at trivial objects
template <typename T>
requires (!HasClearGentlyImpl<T>)
future<> clear_gently(T&) noexcept {
    return make_ready_future<>();
}

} // namespace internal
```

and I'm calling that from the high level `clear_gently` implementations, instead of unqualified `clear_gently`.
it turned out to be pretty neat.

Reply:
Done this way in v6.